R/opts.R

```diff
@@ -1,27 +1,145 @@
-#' "mean", "mean_sd", "sd_mean", "sd_sd", and "max" defining a truncated log
-#' normal (with all parameters except for max defined in logged form).
+#' @param ... Parameters of discretised (upper-)truncated lognormal delay
+#' distributions as a list with the following parameters:
+#' "mean", the mu parameter or mean of the natural logarithm of the delay;
+#' "mean_sd", the standard deviation in the estimate of "mean" parameter
+#' (assumed normally distributed); "sd", the sigma parameter or standard
+#' deviation of the natural logarithm of the delay; "sd_sd", the standard
+#' deviation of the estimate of the "sd" parameter (assumed normally
+#' distributed) sd_sd"; and "max", the maximum delay.
+#' The "mean" parameter is mandatory; if it is the only one given it represents
+#' a fixed delay and must be integer-valued; if "sd" is also given and
+#' greater than 0 this represents a delay distribution and "mean" can be
+#' real-valued. In that case, "max" also needs to be given.
+#' The "mean_sd" and "sd_sd" parameters should be provided to represent
+#' uncertainty in the parameter values of the delay but are optional.
+#' @param fixed Logical, defaults to `FALSE`. Should reporting delays be treated
+#' as coming from fixed (vs uncertain) distributions. Making this simplification
+#' drastically reduces compute requirements.
 #' @seealso convert_to_logmean convert_to_logsd bootstrapped_dist_fit
 #' @return A list summarising the input delay distributions.
 #' @export
 #' @examples
 #' # no delays
 #' delay_opts()
-delay_opts <- function(...) {
+#'
+#' # A single delay that has uncertainty
+#' delay <- list(mean = 1, mean_sd = 0.2, sd = 0.5, sd_sd = 0.1, max = 15)
+#' delay_opts(delay)
+#'
+#' # A single delay where we override the uncertainty assumption
+#' delay_opts(delay, fixed = TRUE)
+#'
+#' # A delay where uncertainty is implict
+#' delay_opts(list(mean = 1, mean_sd = 0, sd = 0.5, sd_sd = 0, max = 15))
+delay_opts <- function(..., fixed = FALSE) {
   delays <- list(...)
   data <- list()
   data$delays <- length(delays)
-  if (data$delays > 0) {
-    delays <- purrr::transpose(delays)
+  for (i in seq_along(delays)) {
+    for (param in c("mean_sd", "sd", "sd_sd")) {
+      if (!(param %in% names(delays[[i]]))) {
+        delays[[i]][[param]] <- 0
+      }
+    }
+    if (fixed) {
+      delays[[i]][["mean_sd"]] <- 0
+      delays[[i]][["sd_sd"]] <- 0
+    }
+    if (delays[[i]]$sd == 0 && delays[[i]]$sd_sd == 0) {
+      if (delays[[i]]$mean_sd > 0) {
+        stop("Error in delay distribution definition: if sd_mean is 0 and ",
+             "sd_sd is 0 then mean_sd must be 0, too.")
+      }
+      if (delays[[i]]$mean %% 1 != 0) {
+        stop("Error: if using a fixed delay it must be integer")
+      }
+    }
+    if (delays[[i]]$mean_sd == 0 &&
+        delays[[i]]$sd == 0 &&
+        delays[[i]]$sd_sd == 0) {
+      delays[[i]]$max <- delays[[i]]$mean
+    }
   }
```

Comment:
```suggestion
  for (d in delays) {
    for (param in c("mean_sd", "sd", "sd_sd")) {
      if (!(param %in% names(d))) {
        d[[param]] <- 0
      }
    }
    if (fixed) {
      d[["mean_sd"]] <- 0
      d[["sd_sd"]] <- 0
    }
    if (d$sd == 0 && d$sd_sd == 0) {
      if (d$mean_sd > 0) {
        stop("Error in delay distribution definition: if sd_mean is 0 and ",
             "sd_sd is 0 then mean_sd must be 0, too.")
      }
      if (d$mean %% 1 != 0) {
        stop("Error: if using a fixed delay it must be integer")
      }
    }
    if (d$mean_sd == 0 &&
        d$sd == 0 &&
        d$sd_sd == 0) {
      d$max <- d$mean
    }
  }
```

You can loop directly on delays here. I find it easier to read but might be subjective.

Reply:
Good point - this has now been replaced by the new `delay_dist` function and [a single `lapply` call](https://github.com/epiforecasts/EpiNow2/blob/ca6d0d71a6aeacde9bb634860539af2e113c91c8/R/opts.R#L103) instead of loop.